### DISCUSSION TITLE
Shader Error Message contains no actual error message...?

### DISCUSSION THREAD

**Discussion by johannesvollmer** [gfx-rs/wgpu#1723] [Q&A] [answered]
Hi! I'm trying to compile the simplest WGSL shader module with WGPU. There's an error but I can't figure out what it is caused by. Googling didn't yield any useful results for me.

### The error:
```
[2021-07-25T11:10:33Z ERROR naga::front::wgsl] |	struct VertexOutput {
[2021-07-25T11:10:33Z ERROR naga::front::wgsl] >	    [[builtin(position)]] clip_position: vec4<f32>;
[2021-07-25T11:10:33Z ERROR naga::front::wgsl] |	             ^
wgpu error: Validation Error

Caused by:
    In Device::create_shader_module
    Failed to parse WGSL


thread 'main' panicked at 'Handling wgpu errors as fatal by default',
```

### Why I can't figure this out on my own
- The error message seems to be cut off? There's no real error message. It's not helpful. Is this a syntax error? Is the `position` attribute not correct? Is there a brace missing? Is this not the place for putting an attribute?
- I disabled validation with `ShaderFlags::empty()`, and still the error message says `Validation Error`
- Compiling the shader panics. Maybe it's cut off because of the way it panics...? I assume this could be changed by replacing the global error handler somewhere? Or just for this one call? What about the existing handler, is it lost? On the surface, it looks like WGPU could just return a `Result::Err(...)`, which would be more idiomatic Rust.

### Background
I'm trying to use `iced_wgpu` with wgpu. As far as I understand, the error is not related to `iced` in any way. I came up with this code when reading [learn wgpu](https://sotrh.github.io/learn-wgpu/).

For completeness, here's the full shader source:
```wgsl

struct VertexOutput {
    [[builtin(position)]] clip_position: vec4<f32>;
};

[[stage(vertex)]]
fn vertex_shader(
    [[builtin(vertex_index)]] in_vertex_index: u32,
) -> VertexOutput
{
    let x = f32(1 - i32(in_vertex_index)) * 0.5;
    let y = f32(i32(in_vertex_index & 1u) * 2 - 1) * 0.5;

    var out: VertexOutput;
    out.clip_position = vec4<f32>(x, y, 0.0, 1.0);
    return out;
}

[[stage(fragment)]]
fn fragment_shader(in: VertexOutput) -> [[location(0)]] vec4<f32> {
    return vec4<f32>(0.3, 0.2, 0.1, 1.0);
}
```

Here's the Rust code I'm trying to compile the shader with:
```rust
let shader_module =
    device.create_shader_module(&wgpu::ShaderModuleDescriptor { // this call panics
        label: None,
        source: wgpu::ShaderSource::Wgsl(Cow::Borrowed(include_str!("simple-shader.wgsl"))),
        flags: ShaderFlags::empty() // previously: ShaderFlags::VALIDATION
    });
```

Note: I will open an issue because the error message should probably be improved. If this is a naga problem, I will also open an Issue there.

**Comment by kvark** [comment]
It does feel like the error is incomplete here, even though it shows the exact location. What happens if you just do
```
git clone https://github.com/gfx-rs/naga
cd naga
cargo run -- <your_file.wgsl>
```

**Reply by johannesvollmer** [reply]
`Validation successful` :)

**Reply by johannesvollmer** [reply]
Browsing the source code, I found some code like this: `#[cfg(feature = "trace")] { /* looks like detailed error code? */ }`. I'll try to find out whether I can activate this feature flag and try to compile the shader again

**Reply by kvark** [reply]
tracing is not for error codes, see https://github.com/gfx-rs/wgpu/wiki/Debugging-wgpu-Applications#tracing-infrastructure

**Reply by johannesvollmer** [reply]
Oh okay. I'll try to run other examples and tests from wgpu to find out whether my machine is the cause

**Reply by johannesvollmer** [reply]
I ran the tests using these commands: 
- `git clone www.github.com/gfx-rs/wgpu` 
- `cd wgpu` 
- `cargo run --bin wgpu-info -- cargo test --no-fail-fast` 

There are some failing tests on my machine. I have an `HP OMEN 15dc` Laptop, unmodified, about 3 years old, the device is `NVIDIA GeForce GTX 1060 on Vulkan` and also I have an `Intel CORE i7 8th Gen`.

Condensed:
- `skybox stdout: 0 outliers over max difference 0`
- `Doc-tests player: process didn't exit successfully (exit code: 0xc0000005, STATUS_ACCESS_VIOLATION)`

### Full Error Log
First failed test:
```
Running unittests (target\debug\examples\skybox-13445620ea830473.exe)

running 4 tests
test skybox_etc2 ... ok
test skybox_astc ... ok
test skybox_bc1 ... ok
test skybox ... FAILED

failures:

---- skybox stdout ----
0 outliers over max difference 0
thread 'skybox' panicked at 'UNEXPECTED TEST PASS: BACKEND', wgpu\examples\skybox\..\..\tests\common\mod.rs:299:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    skybox

test result: FAILED. 3 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 1.57s
```

Second failed test:
```
Doc-tests player

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

error: test failed.

Caused by:
  process didn't exit successfully: `C:\Users\Johannes\Documents\GitHub\wgpu-test-shader-compile\wgpu\target\debug\deps\test-a63410f1b07f0d61.exe` (exit code: 0xc0000005, STATUS_ACCESS_VIOLATION)
  process didn't exit successfully: `C:\Users\Johannes\Documents\GitHub\wgpu-test-shader-compile\wgpu\target\debug\examples\skybox-13445620ea830473.exe` (exit code: 101)
=========== FAILED! NVIDIA GeForce GTX 1060 on Vulkan (1 of 3) in 92.331s ===========
error: process didn't exit successfully: `target\debug\wgpu-info.exe cargo test --no-fail-fast` (exit code: 1)
```

**Reply by johannesvollmer** [reply]
Should I open a separate Issue for this?

**Reply by johannesvollmer** [reply]
I just realized that `iced` probably doesn't use the newest version of `wgpu`. Next, I'll try to compile the same shader with the newest version of wgpu

Edit: The wgpu version that comes with the published `iced` crate is `0.7.0`.  
Edit: So, I got it working by using the git master branch of `iced`, which uses the newest wgpu version. Nevertheless, the shader parsing errors remain pretty unspecific in many situations. Furthermore, there is a bug in the `0.7.0` version which might still exist as an edge case in `0.9.0`, undetected.

**Reply by kvark** [reply]
yes, please file an issue. We aren't supposed to crash during testing.

**Comment by johannesvollmer** [comment, marked as answer]
Using the newest version solved the problem of the (incorrect) error. However, most errors are still quite unspecific. Nevertheless, thanks for the pointers, @kvark 

I would like to open an issue to collect examples where the error messages could be improved. Is this a naga issue or a wgpu issue?

**Reply by kvark** [reply]
If running Naga's "naga-cli" shows proper error (https://github.com/gfx-rs/wgpu/discussions/1723#discussioncomment-1047435), then it's a wgpu issue.

**Reply by johannesvollmer** [reply]
Of course, right! :) Thanks
